src/main/java/com/google/cloud/spanner/pgadapter/utils/MutationWriter.java

```diff
@@ -0,0 +1,203 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.cloud.spanner.pgadapter.utils;
+
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Mutation.WriteBuilder;
+import com.google.cloud.spanner.jdbc.CloudSpannerJdbcConnection;
+import com.google.cloud.spanner.pgadapter.ConnectionHandler;
+import com.google.spanner.v1.TypeCode;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+public class MutationWriter {
+
+  private static final long MUTATION_LIMIT = 20000; // PLACEHOLDER
+  private static final long BATCH_LIMIT = 64000; // PLACEHOLDER
+
+  private boolean hasHeader;
+  private boolean isHeaderParsed;
+  private int mutationCount;
+  private int batchSize;
+  private int rowCount;
+  private List<Mutation> mutations;
+  private String tableName;
+  private Map<String, TypeCode> tableColumns;
+  private CSVFormat format;
+  private FileWriter fileWriter;
+
+  public MutationWriter(
+      String tableName, Map<String, TypeCode> tableColumns, CSVFormat format, boolean hasHeader) {
+    this.mutationCount = 0;
+    this.batchSize = 0;
+    this.hasHeader = hasHeader;
+    this.isHeaderParsed = false;
+    this.tableName = tableName;
+    this.tableColumns = tableColumns;
+    this.format = format;
+    this.mutations = new ArrayList<>();
+  }
+
+  /** @return number of rows copied into Spanner */
+  public int getRowCount() {
+    return this.rowCount;
+  }
+
+  /** Build mutation to add to mutations list with data contained within a CopyData payload */
+  public void buildMutation(ConnectionHandler connectionHandler, byte[] payload) throws Exception {
+    List<CSVRecord> records = parsePayloadData(payload);
+    if (!records.isEmpty()
+        && !payloadFitsInCurrentBatch(records.size() * records.get(0).size(), payload.length)) {
+      writeToSpanner(connectionHandler); // Close out the current batch and reset for the next batch
```

Comment:
This part does not feel right here. The method is called `buildMutation`. We also have a `writeToSpanner` method that is called by another class. That means that a call to `buildMutation` will potentially write some of the mutations to Spanner, but the other class also needs to call `writeToSpanner` to write mutations.

My understanding is that this implementation will not be supporting copy operations that exceed the standard Spanner transaction limits (20,000 mutations). Do we need this call to `writeToSpanner` here if we accept that the copy will fail if it exceeds the limit? If it is not needed, I would suggest that we remove this to avoid any confusion.

Reply:
Removed.  I think we should rollback in the case that it cannot fit inside the mutation limit.

Reply:
Sounds good, thanks.